### ISSUE TITLE
Request for some changes.

### ISSUE THREAD

**Issue by helson73** [OpenNMT/OpenNMT-py#542]
1. Since newest pytorch supports 3D tensors for linear layers, it's better to replace "BottleLinear" to "nn.Linear" for avoiding unnecessary reshaping.
2. When batch is "tokens", calculating loss by shards can't guarantee fixed size of storage. If batch is dynamic, max_generator_batches should be dynamic too.

**Comment by srush** [comment]
Yes, you read my mind! I have implemented them in my code base, but haven't fully tested yet. (I'm mildly worried that 1 will break old model.)

Do you have any cycles to send a PR? I really think 1 is as easy as just changing to nn.Linear and removing from the code. 2 is only slightly harder.

**Comment by helson73** [comment]
@srush 
I will send a PR for 2). 
1) needs more careful consideration. 
Another thing is, current input form of WeightNormConv2d is (batch_size, dim, length, 1), which means each dimension of hidden variables is corresponding to one channel. I suggest change it to (batch_size, 1, length, dim), it should be much faster.

**Comment by srush** [comment]
Sounds good. 

I haven't looked too much at the weight norm work. That was submitted by a contributor.
